Streamline and expand folds

treeowl opened the pull request:
* As discussed in #55, we were recursively constructing `Foldable`
  dictionaries in unordered folds (for the basic queues, though not the
  `Prio` ones). Stop doing that.

* Add strict left unordered folds and monoidal unordered folds.

konsumlamm commented on src/Data/PQueue/Internals/Foldable.hs:
```diff
@@ -0,0 +1,22 @@
+-- | Writing 'Foldable' instances for nonuniform types in the natural manner
```
```suggestion
-- | Writing 'Foldable' instances for non-uniform types (e.g. 'BinomForest') in the natural manner
```
I assume by non-uniform you mean polymorphic recursive?

treeowl replied:
I meant non-regular, sorry. Using non-regular (AKA nested) types tends to require polymorphic recursive functions. I'll fix the terminology.

konsumlamm commented on src/Data/PQueue/Prio/Internals.hs:
```diff
@@ -650,6 +662,17 @@ foldrWithKeyF_ f fCh ts0 z0 = case ts0 of
       f k a (fCh ts (fCh tss z))
 
+-- | Unordered monoidal fold on a binomial forest.
+foldMapWithKeyF_ :: Monoid m => (k -> a -> m) -> (rk k a -> m) -> BinomForest rk k a -> m
```
Why are you defining standalone functions here instead of also uisng a typeclass?

treeowl replied:
Just going by the pattern the module was using before. I don't know which way (if either) performs better. We can evaluate that ... once we have benchmarks. Do you know of any benchmarks out there?

konsumlamm replied:
I don't unfortunately, I've only done a few very basic benchmarks (`insert` and `deleteMin`) a while ago.

I think we should at least be consistent between `Data.PQueue` and `Data.PQueue.Prio`.

treeowl replied:
Fine. I'll go with the class way for now and we can revisit later. I want to pack in a few more features (including deciding how to expose internals) and get a release out.

treeowl commented on src/Data/PQueue/Internals.hs:
```diff
@@ -648,12 +711,28 @@ mapU f (MinQueue n x ts) = MinQueue n (f x) (f <$> ts)
--- | /O(n)/. Unordered left fold on a priority queue.
+-- | /O(n)/. Unordered left fold on a priority queue. This is rarely
+-- what you want; 'foldrU' and 'foldlU'' are more likely to perform
+-- well.
```
True!

konsumlamm commented on src/Data/PQueue/Internals.hs:
```diff
@@ -603,6 +608,63 @@ instance Functor rk => Functor (BinomForest rk) where
   fmap f (Cons t ts) = Cons (fmap f t) (fmap f ts)
 
+instance Foldr Zero where
+  foldr_ _ z ~Zero = z
```
Btw, why are you using a lazy pattern match here?

treeowl replied:
There's no reason to actually match on the `Zero` constructor, since it carries no information. The pattern is only there to make it obvious what is there and why we don't care about it.

konsumlamm commented on src/Data/PQueue/Max.hs:
```diff
@@ -273,10 +277,18 @@ mapU f (MaxQ q) = MaxQ (Min.mapU (\(Down a) -> Down (f a)) q)
+foldMapU :: Monoid m => (a -> m) -> MaxQueue a -> m
+foldMapU f (MaxQ q) = Min.foldMapU (f . unDown) q
+
 -- | /O(n)/. Unordered left fold on a priority queue.
```
You forgot the note about performance.

konsumlamm commented on src/Data/PQueue/Internals.hs:
```diff
@@ -645,12 +708,28 @@ mapU f (MinQueue n x ts) = MinQueue n (f x) (f <$> ts)
+
+-- | /O(n)/. Unordered strict left fold on a priority queue.
+--
+-- @since 1.4.2
```
This seems to be the only file you added `@since` annotations. They should be on all new (exported) functions.

treeowl replied:
Yes, you're right. I'll have to fix that up.

konsumlamm reviewed (commented):
I can only repeat what I said in https://github.com/lspitzner/pqueue/pull/46#pullrequestreview-825645049: please don't just merge non-trivial changes without review.

treeowl commented:
> I can only repeat what I said in [#46 (review)](https://github.com/lspitzner/pqueue/pull/46#pullrequestreview-825645049): please don't just merge non-trivial changes without review.

I thought I'd addressed your concerns. Let's try to keep this friendly.

konsumlamm commented:
> I thought I'd addressed your concerns. Let's try to keep this friendly.

Sorry, I didn't mean to be unfriendly.

The person raising the concerns can usually judge best whether or not they've been addressed, so I think it's useful to await a re-review. Imo, a good rule for (non-controversial) PRs would be that it needs at least one explicit approval (from someone else than the author) to get merged.
